### PR TITLE
Add option to do offload copying automatically

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

**Comment by mvermeulen** [comment]
How do I change the option programatically in Python?

I have verified this branch from Python using the previous python driver I was using.  First data using resnet50 bs=64:
= this branch: 74.544 ms, 859 images/sec
= last measurements (Oct 21): 73.299, 873 images/sec
The previous measurements were based on older version of MIGraphX, so want to use the option to compare using the same MIGraphX build.

File: src/include/migraphx/generate.hpp [hunk 1]
```diff
@@ -78,20 +78,22 @@ struct xorshift_generator
...
 template <class T>
-std::vector<T> generate_tensor_data(const migraphx::shape& s, unsigned long seed = 0)
+auto generate_tensor_data(const migraphx::shape& s, unsigned long seed = 0)
 {
-    std::vector<T> result(s.elements());
-    std::generate(result.begin(), result.end(), xorshf96_generator<T>{seed});
+    // std::vector<T> result(s.elements());
+    auto result = make_shared_array<T>(s.elements());
+    std::generate(result.get(), result.get() + s.elements(), xorshf96_generator<T>{seed});
     // std::generate(result.begin(), result.end(), [&]{ return seed % 7; });
```
Comment (kahmed10) [inline src/include/migraphx/generate.hpp]: comments still needed?

File: src/include/migraphx/generate.hpp [hunk 2]
```diff
@@ -78,20 +78,22 @@ struct xorshift_generator
...
     return result;
 }
 
 template <class T>
-std::vector<T> fill_tensor_data(const migraphx::shape& s, unsigned long value = 0)
+auto fill_tensor_data(const migraphx::shape& s, unsigned long value = 0)
 {
-    std::vector<T> result(s.elements());
-    std::generate(result.begin(), result.end(), [=] { return value; });
+    // std::vector<T> result(s.elements());
```
Comment (kahmed10) [inline src/include/migraphx/generate.hpp]: same as above, commented code

File: src/targets/gpu/hip.cpp [hunk 3]
```diff
@@ -102,29 +132,40 @@ void set_device(std::size_t id)
...
-    std::size_t dst_size = dst.get_shape().bytes();
-    if(src_size > dst_size)
-        MIGRAPHX_THROW("Not enough memory available in destination to do copy");
-    auto status = hipMemcpyAsync(
-        dst.data(), src.data(), src_size, hipMemcpyDeviceToDevice, ctx.get_stream().get());
-    if(status != hipSuccess)
-        MIGRAPHX_THROW("Gpu copy failed: " + hip_error(status));
+    // Workaround: Use contiguous as hip's memcpy is broken
+    device::contiguous(ctx.get_stream().get(), dst, src);
+    // hip_async_copy(ctx, src, dst, hipMemcpyDeviceToDevice);
```
Comment (kahmed10) [inline src/targets/gpu/hip.cpp]: will this be re-enabled in the future, assuming hip_async_copy is working??

Comment (pfultz2) [reply src/targets/gpu/hip.cpp]: Yes, I hope so.

File: src/targets/gpu/include/migraphx/gpu/hip.hpp [hunk 4]
```diff
@@ -77,40 +81,63 @@ struct hip_sync
...
+            gpu_copy(ctx, args[0], result);
+            return result;
+        }
+        copy_from_gpu(ctx, args[0], args[1]);
         return args[1];
     }
-    std::ptrdiff_t output_alias(const std::vector<shape>&) const { return 1; }
+    std::ptrdiff_t output_alias(const std::vector<shape>& args) const
+    {
+        if(args.size() == 1)
```
Comment (kahmed10) [inline src/targets/gpu/include/migraphx/gpu/hip.hpp]: (Optional) this can be rewritten using the ternary operator.

Comment (kahmed10) [reply src/targets/gpu/include/migraphx/gpu/hip.hpp]: Same with L104 above.

File: src/targets/gpu/include/migraphx/gpu/hip.hpp [hunk 5]
```diff
@@ -77,40 +81,63 @@ struct hip_sync
...
+            return -1;
+        return 1;
+    }
 };
 
 struct hip_copy
 {
-    std::string name() const { return "hip_copy"; }
+    std::string name() const { return "hip::copy"; }
```
Comment (scxiao) [inline src/targets/gpu/include/migraphx/gpu/hip.hpp]: I prefer the name as "gpu::hip_copy"

File: src/targets/gpu/include/migraphx/gpu/hip.hpp [hunk 6]
```diff
@@ -124,6 +151,29 @@ struct hip_copy
...
+    shape s;
+    std::string id{};
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.s, "shape"), f(self.id, "id"));
+    }
+
+    std::string name() const { return "hip::hip_load_memory"; }
```
Comment (scxiao) [inline src/targets/gpu/include/migraphx/gpu/hip.hpp]: I prefer the name "gpu::hip_load_memory"

File: src/targets/gpu/target.cpp [hunk 7]
```diff
@@ -72,6 +73,7 @@ std::vector<pass> target::get_passes(migraphx::context& gctx) const
         write_literals{&ctx},
         schedule{gpu::schedule_model{ctx.get_current_device().nstreams()}, not enabled(MIGRAPHX_DISABLE_SCHEDULE_PASS{})},
         memory_coloring{"hip::allocate"},
+        preallocate_param{"scratch", &ctx},
```
Comment (scxiao) [inline src/targets/gpu/target.cpp]: Does this mean scratch memory is always automatically allocated? 
Do we need providing an option for user to allocate the scratch?

**Pull request merged**